chainer/functions/cross_covariance.py

```diff
@@ -0,0 +1,109 @@
+import numpy
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+class CrossCovariance(function.Function):
+
+    """Cross-covariance loss."""
+
+    def __init__(self, use_cudnn=True):
+        self.use_cudnn = use_cudnn
+        self.y_centered = None
+        self.z_centered = None
+        self.covariance = None
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        z_type, y_type = in_types
+
+        type_check.expect(
+            z_type.dtype == numpy.float32,
+            z_type.ndim == 2,
+            y_type.dtype == numpy.float32,
+            y_type.ndim == 2,
+
+            z_type.shape[0] == y_type.shape[0]
+        )
+
+    def check_type_backward(self, in_types, out_types):
+        type_check.expect(
+            in_types.size() == 2,
+            out_types.size() == 1,
+        )
+        y_in_type, z_in_type = in_types
+        out_type, = out_types
+
+        type_check.expect(out_type.dtype == y_in_type.dtype)
+        type_check.expect(out_type.dtype == z_in_type.dtype)
+
+    def forward_cpu(self, inputs):
+        y, z = inputs
+        y_mean = y.mean(axis=0, keepdims=True)
+        z_mean = z.mean(axis=0, keepdims=True)
+        self.y_centered = (y - y_mean)[:, :, None]
+        self.z_centered = (z - z_mean)[:, None, :]
+        self.covariance = (self.y_centered * self.z_centered).mean(axis=0)
+        cost = 0.5 * (self.covariance**2).sum(keepdims=True)
+        return cost.reshape(()),
+
+    def forward_gpu(self, inputs):
+        y, z = inputs
+
+        # Center inputs
+        y_mean = cuda.empty((1, y.shape[1]))
+        z_mean = cuda.empty((1, z.shape[1]))
+        cuda.cumisc.mean(y, axis=0, out=y_mean, keepdims=True)
+        cuda.cumisc.mean(z, axis=0, out=z_mean, keepdims=True)
+        self.y_centered = cuda.cumisc.subtract(y, y_mean)
+        self.z_centered = cuda.cumisc.subtract(z, z_mean)
+
+        # Calculate cross-covariance
+        self.covariance = cuda.empty((y.shape[1], z.shape[1]))
+        cuda.culinalg.add_dot(self.y_centered, self.z_centered, self.covariance,
+                              transa='T', alpha=1./y.shape[0], beta=0.)
+
+        # Calculate cost
+        cost = cuda.cumisc.sum(0.5 * self.covariance**2)
+        return cost,
+
+    def backward_cpu(self, inputs, grad_outputs):
```

Comment:
Fixed. Thanks for the hint, people might want to scale costs. Also modified the tests to check for a scaling. I noticed that this is not done for the other costs, maybe it should be tested there too.